src/compiler/checker.ts

```diff
@@ -37735,17 +37735,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
-    function getReturnTypeOfSingleNonGenericCallSignature(funcType: Type) {
+    function getNonGenericReturnTypeOfSingleCallSignature(funcType: Type) {
         const signature = getSingleCallSignature(funcType);
-        if (signature && !signature.typeParameters) {
-            return getReturnTypeOfSignature(signature);
+        if (signature) {
+            const returnType = getReturnTypeOfSignature(signature);

```

Comment:
I'm concerned that running getReturnType more often will be slower, particularly in cases where the return type *does* contain a type variable and the work has to be discarded.

Reply:
I think (but I'm not sure) that in most cases that return type would still be requested later anyway and the result of this function is cached in `signature.resolvedReturnType`. So hopefully this doesn't introduce any significant work to the compiler here, the only added work should be the `couldContainTypeVariables(returnType)` call and the result of that is also being cached.